### PR TITLE
Step by step guides

### PR DESCRIPTION
**Pull request by awaelchli**

### CODE REVIEW

File: generate.py [hunk 1]
```diff
@@ -107,11 +106,7 @@ def main(
 
     fabric = L.Fabric(accelerator=accelerator, devices=1)
 
-    if dtype is not None:
-        dt = getattr(torch, dtype, None)
-        if not isinstance(dt, torch.dtype):
-            raise ValueError(f"{dtype} is not a valid dtype.")
-        dtype = dt
+    dtype = torch.bfloat16 if torch.cuda.is_bf16_supported() else torch.float32
```
Comment (lantiga) [inline generate.py:109]: Would `torch.float16` be a better option for people running on V100?

The reasoning is that if you don't have `bfloat16`, you likely also don't have a lot of memory.

Comment (awaelchli) [reply generate.py:109]: float16 seems to deteriorate the quality of results too much.

File: docs/inference.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+# Inference
```
Comment (awaelchli) [inline docs/inference.md:1]: The description misses the --dtype bfloat16 because of #146

File: docs/finetune_lora.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+```bash
+python finetune_lora.py
+```
+
+The finetuning requires at least one GPU with ~24 GB memory (GTX 3090).
+This script will save checkpoints periodically to the folder `out/`.
+
+## Testing the model
+
+Coming soon.
```
Comment (awaelchli) [inline docs/finetune_lora.md:29]: TODO: I need to add this after #145

File: docs/download_weights.md
[the remark this answers is not in the document]
Comment (lantiga) [reply docs/download_weights.md:1]: I would keep the README as is and just add links to the relevant howto in each section
As to how to name the folder: maybe howto stands out more

File: docs/download_weights.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+## Convert from HuggingFace
+
+It is also possible to import weights in the format of the HuggingFace [LLaMA](https://huggingface.co/docs/transformers/main/en/model_doc/llama#transformers.LlamaForCausalLM) model.
+Run this script to convert the weights for loading into Lit-LLaMA:
+
+```bash
+python scripts/convert_hf_checkpoint.py \
+    --hf_checkpoint_path path/to/hf/checkpoint/folder \
+    --lit_checkpoint checkpoints/lit-llama.ckpt
```
Comment (lantiga) [inline docs/download_weights.md:40]: We are standardizing around `pth` edtensions in the defaults, let's do the same here.

Comment (lantiga) [reply docs/download_weights.md:40]: This has caused some confusion previously.

File: docs/download_weights.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    --model_size 7B
+```
+
+You can now run [`generate.py` to test the imported weights](download_weights.md).
```
Comment (lantiga) [inline docs/download_weights.md:44]: I don't think this is the right target link

File: docs/finetune_adapter.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+The steps here only need to be done once:
+
+1. Follow the instructions in the [README](README.md) to install the dependencies.
+2. Download and convert the weights and save them in the `./checkpoints` folder as described [here](download_weights.md).
+3. If you want to utilize more than one GPU, you should `pip install deepspeed`.
+4. Download the data and generate the instruction tuning dataset:
+
+   ```bash
+   python scripts/prepare_alpaca.py
```
Comment (lantiga) [inline docs/finetune_adapter.md:17]: I would also mention that if you want to finetune on your own instructions, just create a dataset with the same format (worth showing the template here? or link it?)

File: docs/finetune_adapter.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+   ```
+
+## Running the finetuning
+
+```bash
+python finetune_adapter.py
```
Comment (lantiga) [inline docs/finetune_adapter.md:23]: Our finetuning script don't take any inputs right now.

I think it would be good to show them at least take the path to the checkpoints and the instruction dataset.

File: docs/finetune_adapter.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+The finetuning requires at least one GPU with ~24 GB memory (GTX 3090).
+You can speed up training by setting the `devices` variable in the script to utilize more GPUs if available.
+Depending on the available GPU memory, you can also tune the `micro_batch_size` parameter to utilize the GPU efficiently.
```
Comment (lantiga) [inline docs/finetune_adapter.md:28]: Link to gradient accumulation blog post?

File: docs/finetune_lora.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
+# Finetuning with LoRA
+
+[Low-rank adaption (LoRA)](https://arxiv.org/abs/2106.09685) is a technique to approximate the update to the linear layers in a LLM with a low-rank matrix factorization. This significantly reduces the number of trainable parameters and speeds up training with little impact on the final performance of the model.
+We demonstrate this method by instruction-finetuning LLaMA 7B on the [Alpaca](https://github.com/tatsu-lab/stanford_alpaca) dataset on a **single GTX 3090 (24GB) GPU**.
```
Comment (lantiga) [inline docs/finetune_lora.md:4]: We should maybe say here that we will run training in `bfloat16` automatically if your GPU supports it?

File: docs/finetune_lora.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+## Preparation
+
+The steps here only need to be done once:
+
+1. Follow the instructions in the [README](README.md) to install the dependencies.
+2. Download and convert the weights and save them in the `./checkpoints` folder as described [here](download_weights.md).
+3. Download the data and generate the instruction tuning dataset:
+
+   ```bash
+   python scripts/prepare_alpaca.py
```
Comment (lantiga) [inline docs/finetune_lora.md:15]: Same comment as before

File: docs/finetune_lora.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+   ```
+
+## Running the finetuning
+
+```bash
+python finetune_lora.py
```
Comment (lantiga) [inline docs/finetune_lora.md:21]: Same comment as before

File: docs/inference.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+We demonstrate how to run inference (next token predictin) with the LLaMA base model in the [`generate.py`](generate.py) script:
```
Comment (lantiga) [inline docs/inference.md:3]:
```suggestion
We demonstrate how to run inference (next token prediction) with the LLaMA base model in the [`generate.py`](generate.py) script:
```

File: docs/inference.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+```bash
+python generate.py --prompt "Hello, my name is"
+```
+Output:
+```
+Hello my name is TJ. I have a passion for the outdoors, love hiking and exploring. I also enjoy traveling and learning new things. I especially enjoy long walks, good conversation and a friendly smile.
+```
+
+The script assumes you have downloaded and converted the weights and saved them in the `./checkpoints` folder as described [here](download_weights.md). If you have the weights stored elswehere, you can pass it in as a parameter:
```
Comment (lantiga) [inline docs/inference.md:13]:
```suggestion
The script assumes you have downloaded and converted the weights and saved them in the `./checkpoints` folder as described [here](download_weights.md). If you have the weights stored elsewhere, you can pass the weights file in as a parameter:
```

File: docs/inference.md [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+```bash
+python generate.py --prompt "Hello, my name is" \
+    --checkpoint_path ./checkpoints/lit-llama/7B/state_dict.pth \
+    --tokenizer_path ./checkpoints/lit-llama/tokenizer.model
+```
+
+With the default settings, this will run the 7B model and require ~26 GB of GPU memory (A100 GPU).
```
Comment (lantiga) [inline docs/inference.md:21]: Worth mentioning `bfloat16` at this point (before quantization)

**Review by lantiga — COMMENTED** [review]
Great, made a few comments

File: docs/finetune_adapter.md (hunk 8 above)
Comment (awaelchli) [reply docs/finetune_adapter.md]: I couldn't find a blog post about grad accumulation from us.

File: docs/finetune_lora.md (hunk 11 above)
Comment (awaelchli) [reply docs/finetune_lora.md]: I exposed the args that allow you to pass in your own data and checkpoint location

File: docs/finetune_adapter.md (hunk 7 above)
Comment (awaelchli) [reply docs/finetune_adapter.md]: I exposed the args that allow you to pass in your own data and checkpoint location

File: docs/finetune_adapter.md (hunk 6 above)
Comment (awaelchli) [reply docs/finetune_adapter.md]: I wrote down a step by step instruction with an example of the expected data format

File: docs/finetune_lora.md (hunk 10 above)
Comment (lantiga) [reply docs/finetune_lora.md]: No that wasn't my intention

**Review by lantiga — APPROVED** [review]
Great job @awaelchli , ready to ship!

File: docs/finetune_adapter.md (hunk 8 above)
Comment (lantiga) [reply docs/finetune_adapter.md]: oh maybe it was Seb's. that's ok
